ndk/src/font.rs

```diff
@@ -0,0 +1,457 @@
+    pub const MAX: FontWeight = FontWeight(ffi::AFONT_WEIGHT_MAX as u16);
+}
+
+impl fmt::Display for FontWeight {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match *self {
+            FontWeight::THIN => f.write_str("Thin"),
+            FontWeight::EXTRA_LIGHT => f.write_str("Extra Light (Ultra Light)"),
+            FontWeight::LIGHT => f.write_str("Light"),
+            FontWeight::NORMAL => f.write_str("Normal (Regular)"),
+            FontWeight::MEDIUM => f.write_str("Medium"),
+            FontWeight::SEMI_BOLD => f.write_str("Semi Bold (Demi Bold)"),
+            FontWeight::BOLD => f.write_str("Bold"),
+            FontWeight::EXTRA_BOLD => f.write_str("Extra Bold (Ultra Bold)"),
+            FontWeight::BLACK => f.write_str("Black (Heavy)"),
+            _ => writeln!(f, "{}", self.0),
+        }
+    }
+}
+
+/// The error type returned when an invalid font weight value is passed.
+#[derive(Debug)]
+pub struct FontWeightValueError(());
+
+impl fmt::Display for FontWeightValueError {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.write_str("font weight must be positive and less than or equal to 1000")
+    }
+}
+
+impl std::error::Error for FontWeightValueError {}
+
+impl TryFrom<u16> for FontWeight {
+    type Error = FontWeightValueError;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        FontWeight::new(value)
+    }
+}
+
+/// A 4-byte integer representing an OpenType axis tag.
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct AxisTag(u32);
+
+impl AxisTag {
+    pub const fn from_be(value: u32) -> Result<Self, AxisTagValueError> {
+        // Each byte in a tag must be in the range 0x20 to 0x7E. A space character cannot be
+        // followed by a non-space character. A tag must have one to four non-space characters.
+        // See https://learn.microsoft.com/en-us/typography/opentype/spec/otff#data-types for details.
+
+        let bytes = value.to_be_bytes();
+
+        // Each byte in a tag must be in the range 0x20 to 0x7E.
+        macro_rules! check_byte_range {
+            ($($e:expr)+) => {
+                $(
+                    if !(bytes[$e] as char).is_ascii_graphic() && bytes[$e] != b' ' {
+                        return Err(AxisTagValueError::InvalidCharacter);
+                    }
+                )+
+            };
+        }
+        check_byte_range!(0 1 2 3);
+
+        if bytes[0] == b' ' {
+            return Err(
+                if bytes[1] == b' ' && bytes[2] == b' ' && bytes[3] == b' ' {
+                    // A tag must have one to four non-space characters.
+                    AxisTagValueError::EmptyTag
+                } else {
+                    // A space character cannot be followed by a non-space character.
+                    AxisTagValueError::InvalidSpacePadding
+                },
+            );
+        }
+
+        macro_rules! check_if_valid {
```

Comment:
Can we replace all this macro magic with:

1. A reverse iterator that counts the number of uninterrupted trailing spaces;
2. A `position()` iterator that finds the index of the first space;
3. If there is a space found in 2., and its index is lower than what the number of spaces found in 1. allows, error.

Reply:
We can't, because this function has to be `const fn`. Disregard this suggestion.